File: android/src/main/kotlin/com/superwall/superwallkit_flutter/json/StoreTransactionMapper.kt
import com.superwall.sdk.store.abstractions.transactions.StoreTransaction
import com.superwall.sdk.store.abstractions.transactions.StoreTransactionType
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Date


fun StoreTransaction.pigeonify(): PStoreTransaction {
    val dateFormatter = DateTimeFormatter.ISO_INSTANT

    fun Date.toIsoString(): String? = this.toInstant()
        .atZone(ZoneId.systemDefault())
        .format(dateFormatter)

    return PStoreTransaction(
        configRequestId = configRequestId,
        appSessionId = appSessionId,
        transactionDate = transactionDate?.toIsoString(),
        originalTransactionIdentifier = originalTransactionIdentifier?:"",
        storeTransactionId = storeTransactionId,
        originalTransactionDate = originalTransactionDate?.toIsoString(),
        webOrderLineItemID = webOrderLineItemID,
        appBundleId = appBundleId,
        subscriptionGroupId = subscriptionGroupId,
        isUpgraded = isUpgraded,
        expirationDate = expirationDate?.toIsoString(),
        offerId = offerId,
        revocationDate = revocationDate?.toIsoString()
    )
}

fun StoreTransactionType.pigeonify(): PStoreTransaction {
    val dateFormatter = DateTimeFormatter.ISO_INSTANT

    fun Date.toIsoString(): String? = this.toInstant()
        .atZone(ZoneId.systemDefault())
        .format(dateFormatter)

    return PStoreTransaction(
        configRequestId = "",
        appSessionId = "",
        transactionDate = transactionDate?.toIsoString(),
        originalTransactionIdentifier = originalTransactionIdentifier?:"",
        storeTransactionId = storeTransactionId,
        originalTransactionDate = originalTransactionDate?.toIsoString(),
        webOrderLineItemID = webOrderLineItemID,
        appBundleId = appBundleId,
        subscriptionGroupId = subscriptionGroupId,
        isUpgraded = isUpgraded,
        expirationDate = expirationDate?.toIsoString(),
        offerId = offerId,
        revocationDate = revocationDate?.toIsoString()
    )
}

File: ios/superwallkit_flutter/Sources/superwallkit_flutter/Mappers/StoreTransactionMapper.swift

import Foundation
import SuperwallKit

extension StoreTransaction {
  func pigeonify() -> PStoreTransaction {
    return PStoreTransaction(
      configRequestId: configRequestId,
      appSessionId: appSessionId,
      transactionDate: transactionDate?.isoString,
      originalTransactionIdentifier: originalTransactionIdentifier,
      storeTransactionId: storeTransactionId,
      originalTransactionDate: originalTransactionDate?.isoString,
      webOrderLineItemID: webOrderLineItemID,
      appBundleId: appBundleId,
      subscriptionGroupId: subscriptionGroupId,
      isUpgraded: isUpgraded,
      expirationDate: expirationDate?.isoString,
      offerId: offerId,
      revocationDate: revocationDate?.isoString
    )
  }
}

extension Date {
  static let isoFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.calendar = Calendar(identifier: .iso8601)
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
    return formatter
  }()

  var isoString: String {
    return Self.isoFormatter.string(from: self)
  }
}
